File: FindMinimumInRotatedSortedArray.java
class Solution {
    /* recursive find: O(logN) / O(logN), 100.00% / 60.42%
    public int recFind(int[] nums, int l, int r) {
        int c = (l + r)/2;

        if (c == l || c == r) {
            return nums[l] >  nums[r] ? nums[r] : nums[l];
        }
        
        // just hack it
        if (nums[c] < nums[c-1] && nums[c] < nums[c+1] ) return nums[c];
        
        if (nums[c] > nums[l]) {
            if (nums[c] < nums[r]) {
                return nums[l]; // found
            } else {
                return recFind(nums, c+1, r); // go right
            }
        } else {
            return recFind(nums, l, c-1); // go left
        }
    }
    */
    
    public int findMin(int[] nums) {
        //return recFind(nums, 0, nums.length - 1);
        
        // O(logN) / O(1) solution
        if (nums.length == 1) return nums[0];
        
        int l = 0,
            r = nums.length - 1;
        
        if (nums[l] < nums[r]) return nums[l];
        int c = (l+r)/2;
        while (l <= r) {
            c = (l+r)/2;

            if (nums[c] > nums[c+1]) return nums[c+1];
            if (nums[c-1] > nums[c]) return nums[c];
            
            if (nums[c] > nums[0]) {
                l = c + 1;
            } else {
                r = c-1;
            }
        }
        return -1;
    }
}

File: FindMinimumInRotatedSortedArray.py
class Solution(object):
    # recursive solution
    def findMinRec(self, nums, l, r):
        if l == r or nums[l] < nums[r]:
            return nums[l]
        
        if r-l < 2:
            return min(nums[l], nums[r])
        
        m = (l+r)/2
        if nums[m] > nums[r]:
            return self.findMinRec(nums, m, r)
        else:
            return self.findMinRec(nums, l, m)        
    
    # Runtime: O(logN), Space: O(1)
    def findMinNonRec(self, n):
        r = len(n) - 1
        if r == 0:
            return n[0]
        if r < 0:
            return None
        l = 0
        
        while l + 1 < r:
            m = (l+r)/2
            if (n[m] > n[r]):
                l = m
            else:
                r = m
        
        return min(n[l], n[r])

    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # return self.findMinRec(nums, 0, len(nums) - 1)
        # test cases: [], [1], [1,2,3], [3,1,2], [2,3,1]
        return self.findMinNonRec(nums)
        
